<a href="https://colab.research.google.com/github/rikiindrmwn/all-about-data/blob/main/Klasifikasi%20Kepuasan%20Penumpang%20Pesawat%20Terbang%20Menggunakan%20K-Nearest%20Neighbour%20(KNN)/Klasifikasi_Kepuasan_Penumpang_Pesawat_Terbang_Menggunakan_K_Nearest_Neighbour_(KNN).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
from google.colab import drive

In [2]:
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [3]:
dataset = pd.read_csv('/content/drive/MyDrive/DATA SCIENCE/LEARN PYTHON/KLASIFIKASI KEPUASAN PENUMPANG PESAWAT TERBANG/train.csv', index_col=0)
df = dataset.copy()

In [4]:
df.head()

id  Gender      Customer Type  Age   Type of Travel     Class  \
0   70172    Male     Loyal Customer   13  Personal Travel  Eco Plus   
1    5047    Male  disloyal Customer   25  Business travel  Business   
2  110028  Female     Loyal Customer   26  Business travel  Business   
3   24026  Female     Loyal Customer   25  Business travel  Business   
4  119299    Male     Loyal Customer   61  Business travel  Business   

   Flight Distance  Inflight wifi service  Departure/Arrival time convenient  \
0              460                      3                                  4   
1              235                      3                                  2   
2             1142                      2                                  2   
3              562                      2                                  5   
4              214                      3                                  3   

   Ease of Online booking  ...  Inflight entertainment  On-board service  \
0                       3  ...                       5                 4   
1                       3  ...                       1                 1   
2                       2  ...                       5                 4   
3                       5  ...                       2                 2   
4                       3  ...                       3                 3   

   Leg room service  Baggage handling  Checkin service  Inflight service  \
0                 3                 4                4                 5   
1                 5                 3                1                 4   
2                 3                 4                4                 4   
3                 5                 3                1                 4   
4                 4                 4                3                 3   

   Cleanliness  Departure Delay in Minutes  Arrival Delay in Minutes  \
0            5                          25                      18.0   
1            1                           1                       6.0   
2            5                           0                       0.0   
3            2                          11                       9.0   
4            3                           0                       0.0   

              satisfaction  
0  neutral or dissatisfied  
1  neutral or dissatisfied  
2                satisfied  
3  neutral or dissatisfied  
4                satisfied  

[5 rows x 24 columns]

In [5]:
df.shape

(103904, 24)

In [6]:
df.isna().sum()

id                                     0
Gender                                 0
Customer Type                          0
Age                                    0
Type of Travel                         0
Class                                  0
Flight Distance                        0
Inflight wifi service                  0
Departure/Arrival time convenient      0
Ease of Online booking                 0
Gate location                          0
Food and drink                         0
Online boarding                        0
Seat comfort                           0
Inflight entertainment                 0
On-board service                       0
Leg room service                       0
Baggage handling                       0
Checkin service                        0
Inflight service                       0
Cleanliness                            0
Departure Delay in Minutes             0
Arrival Delay in Minutes             310
satisfaction                           0
dtype: int64

In [7]:
df.dropna(axis=0, inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 103594 entries, 0 to 103903
Data columns (total 24 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   id                                 103594 non-null  int64  
 1   Gender                             103594 non-null  object 
 2   Customer Type                      103594 non-null  object 
 3   Age                                103594 non-null  int64  
 4   Type of Travel                     103594 non-null  object 
 5   Class                              103594 non-null  object 
 6   Flight Distance                    103594 non-null  int64  
 7   Inflight wifi service              103594 non-null  int64  
 8   Departure/Arrival time convenient  103594 non-null  int64  
 9   Ease of Online booking             103594 non-null  int64  
 10  Gate location                      103594 non-null  int64  
 11  Food and drink                     1035

In [8]:
df.drop(columns=['id'], axis=1, inplace=True)

In [9]:
df['Gender'].replace(('Male','Female'), (0,1), inplace=True)
df['Customer Type'].replace(('disloyal Customer','Loyal Customer'), (0,1), inplace=True)
df['Type of Travel'].replace(('Personal Travel', 'Business travel'), (0,1), inplace=True)
df['Class'].replace(('Eco', 'Eco Plus', 'Business'), (0,1,2), inplace=True)
df['satisfaction'].replace(('neutral or dissatisfied', 'satisfied'), (0,1), inplace=True)

In [10]:
df.head()

Gender  Customer Type  Age  Type of Travel  Class  Flight Distance  \
0       0              1   13               0      1              460   
1       0              0   25               1      2              235   
2       1              1   26               1      2             1142   
3       1              1   25               1      2              562   
4       0              1   61               1      2              214   

   Inflight wifi service  Departure/Arrival time convenient  \
0                      3                                  4   
1                      3                                  2   
2                      2                                  2   
3                      2                                  5   
4                      3                                  3   

   Ease of Online booking  Gate location  ...  Inflight entertainment  \
0                       3              1  ...                       5   
1                       3              3  ...                       1   
2                       2              2  ...                       5   
3                       5              5  ...                       2   
4                       3              3  ...                       3   

   On-board service  Leg room service  Baggage handling  Checkin service  \
0                 4                 3                 4                4   
1                 1                 5                 3                1   
2                 4                 3                 4                4   
3                 2                 5                 3                1   
4                 3                 4                 4                3   

   Inflight service  Cleanliness  Departure Delay in Minutes  \
0                 5            5                          25   
1                 4            1                           1   
2                 4            5                           0   
3                 4            2                          11   
4                 3            3                           0   

   Arrival Delay in Minutes  satisfaction  
0                      18.0             0  
1                       6.0             0  
2                       0.0             1  
3                       9.0             0  
4                       0.0             1  

[5 rows x 23 columns]

In [11]:
korelasi_fitur_terhadap_target = df.corr(method='pearson')
korelasi_fitur_terhadap_target['satisfaction'].sort_values(ascending=False)

satisfaction                         1.000000
Online boarding                      0.503447
Class                                0.494545
Type of Travel                       0.448995
Inflight entertainment               0.398203
Seat comfort                         0.349112
On-board service                     0.322450
Leg room service                     0.313182
Cleanliness                          0.305050
Flight Distance                      0.298915
Inflight wifi service                0.284163
Baggage handling                     0.247819
Inflight service                     0.244852
Checkin service                      0.235914
Food and drink                       0.209659
Customer Type                        0.187558
Ease of Online booking               0.171507
Age                                  0.137040
Gate location                        0.000449
Gender                              -0.012356
Departure Delay in Minutes          -0.050515
Departure/Arrival time convenient 

In [12]:
feature_terpilih = korelasi_fitur_terhadap_target[korelasi_fitur_terhadap_target['satisfaction'] > 0.2].index
feature_terpilih

Index(['Type of Travel', 'Class', 'Flight Distance', 'Inflight wifi service',
       'Food and drink', 'Online boarding', 'Seat comfort',
       'Inflight entertainment', 'On-board service', 'Leg room service',
       'Baggage handling', 'Checkin service', 'Inflight service',
       'Cleanliness', 'satisfaction'],
      dtype='object')

In [13]:
df_feature_terpilih = df[feature_terpilih]

X_train_1 = np.array(df_feature_terpilih.iloc[:, :df_feature_terpilih.shape[1]-1])
y_train_1 = np.array(df_feature_terpilih.iloc[:, df_feature_terpilih.shape[1]-1])

X_train_1

array([[   0,    1,  460, ...,    4,    5,    5],
       [   1,    2,  235, ...,    1,    4,    1],
       [   1,    2, 1142, ...,    4,    4,    5],
       ...,
       [   1,    2, 1995, ...,    5,    5,    4],
       [   1,    0, 1000, ...,    5,    4,    1],
       [   1,    2, 1723, ...,    4,    3,    1]])

In [14]:
X = df.iloc[:, :df.shape[1]-1] 
y = df.iloc[:, df.shape[1]-1]

In [15]:
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2

bestfeatures = SelectKBest(score_func=chi2, k=10)
fit = bestfeatures.fit(X,y)

In [16]:
dfscores = pd.DataFrame(fit.scores_)
dfkolom = pd.DataFrame(df.columns)

featurescore = pd.concat([dfkolom, dfscores], axis=1)
featurescore.columns = ['Feature', 'Score']
featurescore.sort_values('Score', ascending=False)

Feature         Score
5                     Flight Distance  7.740606e+06
21           Arrival Delay in Minutes  3.388907e+04
20         Departure Delay in Minutes  2.604190e+04
4                               Class  2.284134e+04
11                    Online boarding  1.470926e+04
2                                 Age  1.128372e+04
13             Inflight entertainment  8.691513e+03
3                      Type of Travel  6.477092e+03
12                       Seat comfort  6.384881e+03
6               Inflight wifi service  5.403193e+03
14                   On-board service  5.284790e+03
15                   Leg room service  5.245905e+03
19                        Cleanliness  5.050652e+03
17                    Checkin service  2.793895e+03
10                     Food and drink  2.513223e+03
16                   Baggage handling  2.443600e+03
18                   Inflight service  2.357584e+03
8              Ease of Online booking  2.162993e+03
1                       Customer Type  6.659929e+02
7   Departure/Arrival time convenient  2.106489e+02
0                              Gender  7.788946e+00
9                       Gate location  1.147346e-02
22                       satisfaction           NaN

In [17]:
feature_terpilih_2 = featurescore.sort_values('Score', ascending=False)['Feature'].head(15).values
feature_terpilih_2

array(['Flight Distance', 'Arrival Delay in Minutes',
       'Departure Delay in Minutes', 'Class', 'Online boarding', 'Age',
       'Inflight entertainment', 'Type of Travel', 'Seat comfort',
       'Inflight wifi service', 'On-board service', 'Leg room service',
       'Cleanliness', 'Checkin service', 'Food and drink'], dtype=object)

In [18]:
df_feature_terpilih_2 = df[feature_terpilih_2]

X_train_2 = np.array(df_feature_terpilih_2.iloc[:, :df_feature_terpilih_2.shape[1]-1])
y_train_2 = np.array(df_feature_terpilih_2.iloc[:, df_feature_terpilih_2.shape[1]-1])

y_train_2

array([5, 1, 5, ..., 4, 1, 1])

In [19]:
from sklearn.neighbors import KNeighborsClassifier

model_1 = KNeighborsClassifier(n_neighbors=5)
model_2 = KNeighborsClassifier(n_neighbors=5)

model_1.fit(X_train_1, y_train_1)
model_2.fit(X_train_2, y_train_2)

KNeighborsClassifier()

In [20]:
dataset_test = pd.read_csv('/content/drive/MyDrive/DATA SCIENCE/LEARN PYTHON/KLASIFIKASI KEPUASAN PENUMPANG PESAWAT TERBANG/test.csv')
df_test = dataset_test.copy()

In [21]:
df_test.dropna(axis=0, inplace=True)
df_test.drop(columns=['id'], axis=1, inplace=True)

df_test['Gender'].replace(('Male','Female'), (0,1), inplace=True)
df_test['Customer Type'].replace(('disloyal Customer','Loyal Customer'), (0,1), inplace=True)
df_test['Type of Travel'].replace(('Personal Travel', 'Business travel'), (0,1), inplace=True)
df_test['Class'].replace(('Eco', 'Eco Plus', 'Business'), (0,1,2), inplace=True)
df_test['satisfaction'].replace(('neutral or dissatisfied', 'satisfied'), (0,1), inplace=True)

In [22]:
df_feature_test_terpilih = df_test[feature_terpilih]

X_test_1 = np.array(df_feature_test_terpilih.iloc[:, :df_feature_test_terpilih.shape[1]-1])
y_test_1 = np.array(df_feature_test_terpilih.iloc[:, df_feature_test_terpilih.shape[1]-1])

X_test_1

array([[   1,    0,  160, ...,    2,    5,    5],
       [   1,    2, 2863, ...,    3,    4,    5],
       [   1,    0,  192, ...,    2,    2,    2],
       ...,
       [   0,    0,  828, ...,    5,    4,    2],
       [   1,    2, 1127, ...,    4,    5,    4],
       [   0,    0,  264, ...,    1,    1,    1]])

In [23]:
df_feature_test_terpilih_2 = df_test[feature_terpilih_2]

X_test_2 = np.array(df_feature_test_terpilih_2.iloc[:, :df_feature_test_terpilih_2.shape[1]-1])
y_test_2 = np.array(df_feature_test_terpilih_2.iloc[:, df_feature_test_terpilih_2.shape[1]-1])

X_test_2

array([[1.600e+02, 4.400e+01, 5.000e+01, ..., 5.000e+00, 5.000e+00,
        2.000e+00],
       [2.863e+03, 0.000e+00, 0.000e+00, ..., 4.000e+00, 5.000e+00,
        3.000e+00],
       [1.920e+02, 0.000e+00, 0.000e+00, ..., 1.000e+00, 2.000e+00,
        2.000e+00],
       ...,
       [8.280e+02, 0.000e+00, 0.000e+00, ..., 3.000e+00, 2.000e+00,
        5.000e+00],
       [1.127e+03, 0.000e+00, 0.000e+00, ..., 2.000e+00, 4.000e+00,
        4.000e+00],
       [2.640e+02, 0.000e+00, 0.000e+00, ..., 2.000e+00, 1.000e+00,
        1.000e+00]])

In [24]:
prediksi_model_1 = model_1.predict(X_test_1)
prediksi_model_2 = model_2.predict(X_test_2)

In [25]:
from sklearn.metrics import accuracy_score

akurasi_model_1 = accuracy_score(y_test_1, prediksi_model_1)
akurasi_model_2 = accuracy_score(y_test_2, prediksi_model_2)

print(f'Akurasi Testing Model 1 : {akurasi_model_1}')
print(f'Akurasi Testing Model 2 : {akurasi_model_2}')

Akurasi Testing Model 1 : 0.8551732128374464
Akurasi Testing Model 2 : 0.30402039161163247


In [26]:
import joblib

joblib.dump(model_1, 'model_knn')

['model_knn']

# Penjelasan Model


1.   Notebook ini berusaha untuk membuat model yang dapat mengklasifikasikan kepuasan penumpang pesat terbang dengan menggunakan algoritma K-Nearest Neighbour (KNN)
2.   Terdapat 2 (dua) teknik seleksi fitur yang dicoba yaitu seleksi fitur berdasarkan hubungannya dengan target dan seleksi fitur menggunakan perhitungan chi square. Kedua hasil seleksi fitur tersebut, digunakan sebagai data training pada model yang berbeda yaitu model_1 untuk seleksi fitur berdasarkan hubungannya terhadap data target dan model_2 untuk seleksi fitur menggunakan chi square
3.   Berdasarkan hasil akurasi testing yang diperoleh, maka disimpulkan bahwa data fitur yang digunakan pada model_1, lebih valid jika dibandingkan dengan fitur yang digunakan pada model_2, karena model_1 menghasilkan akurasi yang lebih baik yaitu 0.8551732128374464, sedangkan model_2 hanya menghasilkan akurasi sebesar 0.3040203916116324.
4.   Untuk itu model yang akan digunakan adalah model_1.

